module/zfs/dsl_prop.c

```diff
@@ -1123,10 +1125,11 @@ dsl_prop_get_all_ds(dsl_dataset_t *ds, nvlist_t **nvp,
 		dsl_dir_name(dd, setpoint);
 		err = dsl_prop_get_all_impl(mos,
 		    dsl_dir_phys(dd)->dd_props_zapobj, setpoint, flags, *nvp);
-		if (err)
-			break;
+		if (err) {
+			nvlist_free(*nvp);
+			return (err);
+		}
 	}
-out:
```

Comment:
@behlendorf  Thank you for your help and guide，benefit a lot from. troube you to review again!